src/learning/lda.rs

```diff
@@ -0,0 +1,250 @@
+            for _ in 0..inputs.1 {
+                for (document, row) in matrix.row_iter().enumerate() {
+                    word_index = 0;
+                    let mut document_topics = unsafe{topics.get_unchecked_mut(document)};
+                    for (word, word_count) in row.iter().enumerate() {
+                        for _ in 0..*word_count {
+                            // Remove the current word from the counts
+                            let mut topic = *unsafe{document_topics.get_unchecked(word_index)};
+                            result.document_topic_count[[document, topic]] -= 1.0;
+                            result.topic_total_by_document[document] -= 1.0;
+                            result.topic_word_count[[topic, word]] -= 1.0;
+                            result.word_total_by_topic[topic] -= 1.0;
+
+                            // Find the probability of that word being a part of each topic
+                            // based on the other words in the present document
+                            let probability = self.conditional_distribution(&result, document, word, vocab_count, topic_count);
+
+                            // Aandomly assign a new topic based on the probabilities from above
+                            topic = choose_from(probability);
+
+                            // Update the counts with the new topic
+                            result.document_topic_count[[document, topic]] += 1.0;
+                            result.topic_total_by_document[document] += 1.0;
+                            result.topic_word_count[[topic, word]] += 1.0;
+                            result.word_total_by_topic[topic] += 1.0;
+                            document_topics[word_index] = topic;
+                            word_index += 1;
+                        }
+                    }
+                }
+            }
+            Ok(result)
+        }
+
+        /// Train the model using inputs  Does nothing on this model.
+        fn train(&mut self, _: &(Matrix<usize>, usize)) -> LearningResult<()> {
+            Ok(())
+        }
+}
+
+/// This function models sampling from the categorical distribution.
+/// That is, given a series of discrete categories, each with different probability of occurring,
+/// this function will choose a category according to their probabilities.
+/// The sum of the probabilities must be 1, but since this is only used internally,
+/// there is no need to verify that this is true.
+fn choose_from(probability: Vector<f64>) -> usize {
+    let mut rng = thread_rng();
+    let selection:f64 = rng.gen_range(0.0, 1.0);
+    let mut total:f64 = 0.0;
+    for (index, p) in probability.iter().enumerate() {
+        total += *p;
+        if total >= selection {
+            return index;
+        }
+    }
+    return probability.size() - 1;
+}
+
+#[cfg(test)]
+mod test {
+    use super::{LDAResult, LDA};
+    use linalg::{Matrix, Vector};
+    #[test]
+    fn test_conditional_distribution() {
```

Comment:
> This test is based on values I calculated basically by hand, and verified by a python implementation.

I think this is fine but you should note this somewhere near the test with a reference to any implementation you used if possible. It makes them a little more convincing to people reading over the code.

>  unless we want to make it possible to pass a source of randomness into a model

This is actually a good idea. See #138 . The support isn't there in the library yet but if you can get something working in the mean time it would be helpful for later. If we want to make changes later we'd like to be able to verify that the algorithm is still doing (close-to) the same thing.